[1.19.x] Document Forge extensions of the model JSON format (#493)
MinecraftForge/Documentation: docs/rendering/modelextensions/transforms.md

Root Transforms
===============

> Adding the `transform` entry at the top level of a model JSON allows specifying a transformation that will be baked into the vertex data of the quads. These transforms apply before the transforms specified in the [blockstate][blockstate] file of a block model and the [display transforms][displaytransform] of an item model.

Reviewer: You can simplify `[blockstate][blockstate]` to just `[blockstate]`.

Also, you should talk about transformations as to what they are applied to. For example, the line is ambiguous for whether the transformation is applied to the model as a whole or onto an individual face.

Author: Made it clearer that it applies to the full model

The root transforms can be specified in four formats:

- A string referring to a default value (currently only supports the value `"identity"`)
- A raw transformation matrix in the form of a nested JSON array with the last row ommited (3*4 matrix, row major order)
- A JSON object containing a singular `matrix` entry with the same format as above

> - A JSON object containing any combination of the following optional entries:

Reviewer: Provide examples with comments. You could infer what this means, but always assume that they don't have the relevant knowledge or context.

Author: I've separated the elementwise specification into a separate section to provide some examples

    - `origin`: origin point used for the rotations and scaling, defaults to `opposing-corner`
        - Array with 3 floating point values for x, y, z

>         - One of the default values: `"corner"` (0, 0, 0), `"center"` (.5, .5, .5) or `"opposing-corner"` (1, 1, 1)

Reviewer: Unclear how the interplay between the two options work. From what I gather, these can be combined together, but it seems as though that it should be one or the other, not both.

Author: Made it a bit clearer that it's an either-or

    - `translation`: relative translation, defaults to (0, 0, 0)
        - Array with 3 floating point values for x, y, z
    - `rotation` or `left_rotation`: rotation around the translated origin to be applied before scaling, defaults to no rotation
        - Single JSON object with a single axis -> rotation degree mapping
        - Array of an arbitrary amount of JSON objects with the above format
        - Array of 3 floating point values specifying the rotation in degrees around each axis
        - Array of 4 floating point values specifying a quaternion directly
    - `scale`: scale relative to the translated origin, defaults to (1, 1, 1)
        - Single floating point value
        - Array with 3 floating point values for x, y, z

>     - `right_rotation` or `post_rotation`: rotation around the translated origin to be applied after scaling, defaults to no rotation

Reviewer: Are the transformations performed in the order provided?

Author: Added an explicit comment about the order of operation

Root Transforms
===============

Adding the `transform` entry at the top level of a model JSON allows specifying a transformation that will be baked into the vertex data of all quads of the model. These transforms apply before the transforms specified in the [blockstate] file of a block model and the [display transforms][displaytransform] of an item model.

The root transforms can be specified in four formats:

1. A string referring to a default value (currently only supports the value `"identity"`)

> 2. A raw transformation matrix in the form of a nested JSON array with the last row ommited (3*4 matrix, row major order)

Reviewer 2: ```suggestion
2. A raw transformation matrix in the form of a nested JSON array with the last row omitted (3*4 matrix, row major order)
```

> 3. A JSON object containing a singular `matrix` entry with the same format as above

Reviewer: I would still like an example for these as it not generally clear how this would look:

```js
{
    "transform": [
       [4, 5, 6], // What do these numbers represent?
       // ...
    ]
}
```

Author: I've added an example purely to demonstrate the expected structure. At the moment, I can't explain what the individual values in the matrix represent as I am failing to even find a simple matrix that works at all. While searching one, I may have found a severe bug in this implementation (apart from two of the transform formats turning out to not be usable at the moment) but I'll need to confirm that with further tests.

4. A JSON object containing any combination of the following optional entries:
    - `origin`: origin point used for the rotations and scaling
    - `translation`: relative translation
    - `rotation` or `left_rotation`: rotation around the translated origin to be applied before scaling
    - `scale`: scale relative to the translated origin
    - `right_rotation` or `post_rotation`: rotation around the translated origin to be applied after scaling

> Elementwise specification

Reviewer 2: ```suggestion
Element wise specification
```

Author: Opted for a dash instead of a space

-------------------------

If the transformation is specified as a combination of the entries mentioned in option 4, these entries will be applied in the order of `translation`, `left_rotation`, `scale`, `right_rotation`.  
The transformation is moved to the specified origin as a last step.

```js
{
    "transform": {
        "origin": "center",
        "translation": [ 0, 0.5, 0 ],
        "rotation": { "y": 45 }
    },
    // ...
}
```

The elements are expected to be defined as follows:

### Origin

> The origin can be specified either as an array of 3 floating point values representing a three dimensional vector: `[ x, y, z ]` or as one of the three default values:

Reviewer 2: ```suggestion
The origin can be specified either as an array of 3 floating point values representing a three-dimensional vector: `[ x, y, z ]` or as one of the three default values:
```

- `"corner"` (0, 0, 0)
- `"center"` (.5, .5, .5)
- `"opposing-corner"` (1, 1, 1)

If the origin is not specified, it defaults to `"opposing-corner"`.

### Translation

The translation must be specified as an array of 3 floating point values representing a three dimensional vector: `[ x, y, z ]` and defaults to (0, 0, 0) if not present.

### Left and Right Rotation

The rotations can be specified in any one of the following four ways:

> - Single JSON object with a single axis -> rotation degree mapping: `{ "X": 90 }`

Reviewer 2: This could use an actual arrow, such as `⇒`

> - Single JSON object with a single axis => rotation degree mapping: `{ "X": 90 }`

Reviewer: Is this supposed to be capitalized `X`

Author: Nice catch, supposed to be lowercase

- Array of an arbitrary amount of JSON objects with the above format (applied in the order they are specified in): `[ { "x": 90 }, { "y": 45 }, { "x": -22.5 } ]`
- Array of 3 floating point values specifying the rotation in degrees around each axis: `[ 90, 180, 45 ]`
- Array of 4 floating point values specifying a quaternion directly: `[ 0.38268346, 0, 0, 0.9238795 ]` (example equals 45 degrees around the X axis)

If the respective rotation is not specified, it will default to no rotation.

### Scale

> The scale must be specified as an array of 3 floating point values representing a three dimensional vector: `[ x, y, z ]` and defaults to (1, 1, 1) if not present.

Reviewer 2: ```suggestion
The scale must be specified as an array of 3 floating point values representing a three-dimensional vector: `[ x, y, z ]` and defaults to (1, 1, 1) if not present.
```

Root Transforms
===============

Adding the `transform` entry at the top level of a model JSON allows specifying a transformation that will be baked into the vertex data of all quads of the model. These transforms apply before the transforms specified in the [blockstate] file of a block model and the [display transforms][displaytransform] of an item model.

> Custom model loaders are given these transforms combined with the [blockstate] transforms or [display transforms][displaytransform] in the form of a `ModelState` in `IGeometryLoader#bake()` and are free to ignore them.

Reviewer 3: ```suggestion
Adding the `transform` entry at the top level of a model JSON suggests to the loader that a transformation should be applied to all geometry right before the rotations in the [blockstate] file in the case of a block model, and before the [display transforms][displaytransform] in the case of an item model.

Custom model loaders may ignore this field entirely.
```

The root transforms can be specified in two formats:

1. A JSON object containing a singular `matrix` entry containing a raw transformation matrix in the form of a nested JSON array with the last row omitted (3*4 matrix, row major order). Example to demonstrate the structure:
   ```js
   "transform": {
       "matrix": [
           [ 0, 0, 0, 0 ],
           [ 0, 0, 0, 0 ],
           [ 0, 0, 0, 0 ]
       ]
   }

>    ```

Reviewer: Should be indented 4 spaces

2. A JSON object containing any combination of the following optional entries:
    - `origin`: origin point used for the rotations and scaling
    - `translation`: relative translation
    - `rotation` or `left_rotation`: rotation around the translated origin to be applied before scaling
    - `scale`: scale relative to the translated origin

>     - `right_rotation` or `post_rotation`: rotation around the translated origin to be applied after scaling

Reviewer: Use `*` instead of `-`
